### ISSUE TITLE
Where does image embeddings come from in Layoutlm funsd task?

### ISSUE THREAD

**Issue by akshat-khare** [NielsRogge/Transformers-Tutorials#3]
Hello @NielsRogge . Thanks a lot for the wonderful tutorial of Fine_tuning_LayoutlmfortokenClassification_on_Funsd.
I have been able to successfully run it on my system. However, I have a basic practical doubt. In the paper of Layoutlm, the authors point out that Layoutlm uses small image clips for each bounding box for training purposes too. 
Page 2 of the paper
```
meanwhile the image embedding can capture some appearance features such as font directions, types, and colors.
```
But these small crops of each bounding box was not made in any of the code you did. Is that task is done while doing forward pass in the below code you uploaded?
```      
outputs = model(input_ids=input_ids, bbox=bbox, attention_mask=attention_mask, token_type_ids=token_type_ids,
                      labels=labels)
```

Basically, I cannot find any part of the code that calculated image embedding from bounding boxes while the paper claims it does. Can you help me find the part of code that does that as I think open-source code which they have uploaded might not have that code. Thanks a lot, man. Looking forward to your response.

**Comment by akshat-khare** [comment]
Hello @NielsRogge , I found that LayoutLm has not included image embedding in the pipeline as is not trained and we can use the any image embedding by appending it to Layoutlm for downstream tasks. They have left that task to people who would use LayoutLm for downstream tasks (people like us). Can you (us) do a tutorial on that too? It would be very helpful as there are none so far which show how we can use image embedding too.
Link to clarification by authors: https://github.com/microsoft/unilm/issues/163
Link to how we can do it: https://github.com/microsoft/unilm/issues/249
Topics talking about image embedding: https://github.com/microsoft/unilm/issues?q=is%3Aissue+is%3Aopen+image+embedding
Thanks

**Comment by NielsRogge** [comment]
There was a lot of interest in this, so I've created a notebook that illustrates how to add image embeddings to LayoutLM: https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Add_image_embeddings_to_LayoutLM.ipynb

**Comment by akshat-khare** [comment]
Thanks a lot @NielsRogge . You are a savior. I highly admire your quick reply. Keep up the good work.

**Comment by akshat-khare** [comment]
@NielsRogge Can you also do a tutorial on how to use Layoutlm embeddings with custom top layers for classification or labeling or any other general task? I guess this would complete the set of tutorials. Currently the tutorials don't have much explanation on how to customise the models to user's need. Thanks a lot for the amazing work man!

**Comment by cydal** [comment]
Thanks for this great tutorial Niel! For inference, you passed the image through FUNSD class to generate the required tokens including a label token. This requires the image's json file. How could this be used with a totally new image that has no  annotation file?

**Comment by NielsRogge** [comment]
Hi @cydal,

I've made a new notebook that shows inference on new images when no label information is accessible. You can find it [here](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLMv2/FUNSD/True_inference_with_LayoutLMv2ForTokenClassification_%2B_Gradio_demo.ipynb).

**Comment by cydal** [comment]
> Hi @cydal,
> 
> I've made a new notebook that shows inference on new images when no label information is accessible. You can find it [here](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLMv2/FUNSD/True_inference_with_LayoutLMv2ForTokenClassification_%2B_Gradio_demo.ipynb).

Thank you for the lightning fast reply. And thank you so much the link. 
The tutorial I was referring to is for the ["adding image embeddings to LayoutLM"](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Add_image_embeddings_to_LayoutLM.ipynb). 
I understand if you don't have something on that already. Could you provide any advice on how to run inference on that without label information?

**Comment by cydal** [comment]
@NielsRogge thanks again for the response. For the notebook for layoutlm with image embeddings, I implemented a new funsd class that uses OCR to extract the word and bbox information. The issue I'm having now is that the performance is poor when I extract manually compared to when I use yours that extracts information from the annotation file. I'm including the model definition below. Can you spot any issues with it? 

`
class inferFUNSDDataset(Dataset):
    """LayoutLM dataset with visual features."""

    def __init__(self, image_file_names, tokenizer, max_length, target_size, train=True):
        self.image_file_names = image_file_names
        self.tokenizer = tokenizer
        self.max_seq_length = max_length
        self.target_size = target_size
        self.pad_token_box = [0, 0, 0, 0]
        self.train = train

    def __len__(self):
        return len(self.image_file_names)

    def __getitem__(self, idx):

        # first, take an image
        item = self.image_file_names[idx]

        original_image = Image.open(item).convert("RGB")
        # resize to target size (to be provided to the pre-trained backbone)
        resized_image = original_image.resize((self.target_size, self.target_size))
        
        # first, read in annotations at word-level (words, bounding boxes, labels)
        width, height = original_image.size
        w_scale = 1000/width
        h_scale = 1000/height

        ocr_df = pytesseract.image_to_data(original_image, output_type='data.frame') \
                    
        ocr_df = ocr_df.dropna() \
                      .assign(left_scaled = ocr_df.left*w_scale,
                              width_scaled = ocr_df.width*w_scale,
                              top_scaled = ocr_df.top*h_scale,
                              height_scaled = ocr_df.height*h_scale,
                              right_scaled = lambda x: x.left_scaled + x.width_scaled,
                              bottom_scaled = lambda x: x.top_scaled + x.height_scaled)

        float_cols = ocr_df.select_dtypes('float').columns
        ocr_df[float_cols] = ocr_df[float_cols].round(0).astype(int)
        ocr_df = ocr_df.replace(r'^\s*$', np.nan, regex=True)
        ocr_df = ocr_df.dropna().reset_index(drop=True)

        words = list(ocr_df.text)
        coordinates = ocr_df[['left', 'top', 'width', 'height']]
        unnormalized_word_boxes = []
        for idx, row in coordinates.iterrows():
          x, y, w, h = tuple(row) # the row comes in (left, top, width, height) format
          actual_box = [x, y, x+w, y+h] # we turn it into (left, top, left+widght, top+height) to get the actual box 
          unnormalized_word_boxes.append(actual_box)



        normalized_word_boxes = [normalize_box(bbox, width, height) for bbox in unnormalized_word_boxes]
        assert len(words) == len(normalized_word_boxes)


        # next, transform to token-level (input_ids, attention_mask, token_type_ids, bbox, labels)
        token_boxes = []
        unnormalized_token_boxes = []
        token_labels = []
        for word, unnormalized_box, box in zip(words, unnormalized_word_boxes, normalized_word_boxes):
            word_tokens = self.tokenizer.tokenize(word)
            unnormalized_token_boxes.extend(unnormalized_box for _ in range(len(word_tokens)))
            token_boxes.extend(box for _ in range(len(word_tokens)))
            # label first token as B-label (beginning), label all remaining tokens as I-label (inside)


                
        
        # Truncation of token_boxes + token_labels
        special_tokens_count = 2 
        if len(token_boxes) > self.max_seq_length - special_tokens_count:
            token_boxes = token_boxes[: (self.max_seq_length - special_tokens_count)]
            unnormalized_token_boxes = unnormalized_token_boxes[: (self.max_seq_length - special_tokens_count)]
            #token_labels = token_labels[: (self.max_seq_length - special_tokens_count)]
        
        # add bounding boxes and labels of cls + sep tokens
        token_boxes = [[0, 0, 0, 0]] + token_boxes + [[1000, 1000, 1000, 1000]]
        unnormalized_token_boxes = [[0, 0, 0, 0]] + unnormalized_token_boxes + [[1000, 1000, 1000, 1000]]
        #token_labels = [-100] + token_labels + [-100]
        
        encoding = self.tokenizer(' '.join(words), padding='max_length', truncation=True)
        # Padding of token_boxes up the bounding boxes to the sequence length.
        input_ids = self.tokenizer(' '.join(words), truncation=True)["input_ids"]
        padding_length = self.max_seq_length - len(input_ids)
        token_boxes += [self.pad_token_box] * padding_length
        unnormalized_token_boxes += [self.pad_token_box] * padding_length
        #token_labels += [-100] * padding_length
        encoding['bbox'] = token_boxes
        #encoding['labels'] = token_labels

        assert len(encoding['input_ids']) == self.max_seq_length
        assert len(encoding['attention_mask']) == self.max_seq_length
        assert len(encoding['token_type_ids']) == self.max_seq_length
        assert len(encoding['bbox']) == self.max_seq_length
        #assert len(encoding['labels']) == self.max_seq_length

        encoding['resized_image'] = ToTensor()(resized_image)
        # rescale and align the bounding boxes to match the resized image size (typically 224x224) 
        encoding['resized_and_aligned_bounding_boxes'] = [resize_and_align_bounding_box(bbox, original_image, self.target_size) 
                                                          for bbox in unnormalized_token_boxes]

        encoding['unnormalized_token_boxes'] = unnormalized_token_boxes
        
        # finally, convert everything to PyTorch tensors 
        for k,v in encoding.items():
            encoding[k] = torch.as_tensor(encoding[k])
        
        return encoding`
